File: pain_2plane_split/apply_curation.m
% apply_curation.m  -  re-derive traces for the hand-curated footprint sets.
%
% Run apply_curation.py first; it writes curated_S.mat per plane.
%
% WHY THIS GOES BACK THROUGH EXTRACT
%   Two of the curation operations change what a trace means, and a plain
%   footprint-weighted average would get both wrong:
%     * pain plane B #11 was split into two somata. Their footprints touch,
%       so a weighted average of each would carry a large fraction of the
%       other's signal. EXTRACT's T-step solves all sources jointly and
%       demixes them.
%     * open-field plane B gained three footprints (N3, N4, N5) seeded as
%       gaussian disks. They need the S-step to take their real shape from
%       the movie, and N4/N5 sit within ~25 px of #4, so again they have to
%       be solved together rather than one at a time.
%
%   The motion-corrected movie is reused from <plane>_1.h5 rather than
%   recomputed (skip_steps_1_2), so this costs seconds of registration, not
%   minutes. Cell finding is off and every quality gate is disabled: the keep
%   or discard calls were made by eye, and re-applying the automatic filters
%   would silently undo them.
%
% ISOLATION - writes only into <plane folder>\curated\, never over
% final_analysis_results.mat and never into the other agent's output\.

clear; clc;
here = fileparts(mfilename('fullpath'));
addpath(here);
mini2p_toolbox_paths();

SESSIONS = { ...
  'openfield', 'D:\20260911_CEANTSR1_65_15_openfield(100_50um)_555mi_2026-09-11_15-27-52'; ...
  'pain',      'D:\20260911_CEANTSR1_65_15_pain_ 10min_pin_heat_2026-09-11_15-56-48' };
PLANES = {'A', 'B'};
% ONLY = {} runs everything; otherwise list 'session plane' keys, e.g.
% {'openfield B', 'pain B'}, to redo just those. SKIP_DONE leaves a plane
% alone if its curated result is already on disk.
ONLY = {};
SKIP_DONE = false;

o = struct();
o.ca_frame_rate_hz = 4.6054;           % measured from the Slice timestamps
o.extract_mode = 'curated';
o.trace_on_unbinned = true;
o.skip_steps_1_2 = true;               % reuse <plane>_1.h5
o.use_zscore_before_extract = false;
o.extract_preset = 'low_snr';
o.avg_cell_radius = 8;
o.skip_morph_filter = true;
o.n_frames_check = 0;
o.skip_video_01 = true;
o.skip_video_03 = true;
o.save_trace_files = true;

% No parpool here on purpose. The curated pass is one alternating update on
% an already motion-corrected movie, so workers buy nothing, and asking for
% a pool while the other agent's MATLAB holds one made this script exit
% silently with no output at all.

summary = {};
for s = 1:size(SESSIONS, 1)
    key = SESSIONS{s, 1};
    sdir = SESSIONS{s, 2};
    for k = 1:numel(PLANES)
        pl = PLANES{k};
        tagkey = [key, ' ', pl];
        if ~isempty(ONLY) && ~any(strcmpi(tagkey, ONLY)), continue; end
        pdir = fullfile(sdir, 'output_split', ['plane_', pl]);
        if SKIP_DONE && isfile(fullfile(pdir, 'curated', ...
                                        'final_analysis_results.mat'))
            fprintf('  %s : already done, skipped\n', tagkey);
            continue;
        end
        tif = fullfile(sdir, 'output_split', ['plane_', pl, '.tif']);
        sfile = fullfile(pdir, 'curated', 'curated_S.mat');
        if ~isfile(sfile)
            warning('missing %s - run apply_curation.py first', sfile);
            continue;
        end
        % step 2 derives <curr_header>_1.h5 from the tif name and looks for
        % it in the output folder, so the curated results go in a subfolder
        % and the h5 is linked in rather than copied (2.7 GB each)
        cout = fullfile(pdir, 'curated');
        h5_src = fullfile(pdir, ['plane_', pl, '_1.h5']);
        h5_dst = fullfile(cout, ['plane_', pl, '_1.h5']);
        if ~isfile(h5_dst)
            if ~isfile(h5_src)
                warning('missing %s', h5_src); continue;
            end
            [ok, msg] = system(sprintf('mklink /H "%s" "%s"', h5_dst, h5_src));
            if ok ~= 0 || ~isfile(h5_dst)
                warning(['hard link failed (%s); copying 2.7 GB instead. ' ...
                         'Delete %s when done.'], strtrim(msg), h5_dst);
                copyfile(h5_src, h5_dst);
            end
        end

        oo = o;
        oo.curated_S_file = sfile;
        fprintf('\n===== curated: %s plane %s =====\n', key, pl);
        t0 = tic;
        try
            pain_2plane_step2_extract(tif, cout, oo);
            f = fullfile(cout, 'final_analysis_results.mat');
            n = NaN;
            if isfile(f)
                S = load(f, 'output');
                n = size(S.output.spatial_weights, 3);
            end
            summary{end+1} = sprintf('  %-10s plane %s : %3d cells, %.0f s', ...
                                     key, pl, n, toc(t0)); %#ok<SAGROW>
        catch ME
            fprintf(2, '  FAILED: %s\n', ME.message);
            for e = 1:numel(ME.stack)
                fprintf(2, '      %s line %d\n', ME.stack(e).name, ...
                        ME.stack(e).line);
            end
            summary{end+1} = sprintf('  %-10s plane %s : FAILED (%s)', ...
                                     key, pl, ME.message); %#ok<SAGROW>
        end
    end
end

fprintf('\n==================== summary ====================\n');
for i = 1:numel(summary), disp(summary{i}); end

File: pain_2plane_split/apply_curation.py
"""apply_curation.py  -  build the curated footprint set from Hansol's decisions.

This only builds the SPATIAL footprints. Traces are re-derived afterwards by
apply_curation.m, which runs EXTRACT's alternating estimation with these
footprints as S_init. That matters for the split and the added cells: a plain
footprint-weighted average of two adjacent somata cross-contaminates them,
whereas EXTRACT's T-step demixes overlapping sources. Nothing here touches
the original final_analysis_results.mat.

OPERATIONS
  discard  drop the footprint entirely
  trim     keep only the largest connected blob of the footprint
  split    make one footprint per connected blob, labelled 11a, 11b, ...
  add      seed a new footprint as a gaussian disk at a given centroid;
           EXTRACT's S-step then shapes it against the movie

DECISIONS, 2026-09-11, from the curation sheets
  OF-A  discard 7, 13
  OF-B  discard 9; trim 14 to its large blob; add N3, N4, N5 - three somata
        EXTRACT missed in the diagonal chain #13 -> N3 -> N4 -> #4 -> N5 -> #2.
        N4 and N5 are the "next to 4" and the clear one between #4 and #2;
        N3 is the third. N3 and N4 are faint (band-pass peak 1.6 and 2.2
        against 3.5 for N5 and 17.2 for #4), so whether they are dim cells or
        neuropil is left for the QC on their re-extracted traces to settle.
  PA-A  discard 9 only - its contour is empty in both the mean and max images
        (anat 0.001, the lowest trace-versus-pixels agreement in the plane at
        0.64). Every other candidate kept, including #6 and #12, which are
        elongated (ecc 0.934, 0.941) and may be fibres rather than somata.
  PA-B  split 11 into two cells; keep everything else

OUTPUTS  ->  <session>\\output_split\\plane_<X>\\curated\\
  curated_S.mat        S (h*w x k), fov_size, labels - input to apply_curation.m
  curated_labels.csv   new index -> label, origin, operation, area

USAGE
  python apply_curation.py
"""
from __future__ import annotations

import os

import cv2
import h5py
import numpy as np
import pandas as pd
from scipy.io import savemat

SESSIONS = {
    "openfield": "D:/20260911_CEANTSR1_65_15_openfield(100_50um)_555mi_"
                 "2026-09-11_15-27-52",
    "pain": "D:/20260911_CEANTSR1_65_15_pain_ 10min_pin_heat_"
            "2026-09-11_15-56-48",
}
REL = 0.3          # blob level, as a fraction of the footprint's own peak
MIN_BLOB = 20      # px; smaller pieces are fragments, not a second soma
SEED_RADIUS = 8.0  # measured median cell radius

DECISIONS = {
    ("openfield", "A"): dict(discard=[7, 13]),
    ("openfield", "B"): dict(discard=[9], trim=[14],
                             add=[("N3", 281.6, 314.6),
                                  ("N4", 281.8, 343.1),
                                  ("N5", 296.7, 388.2)]),
    ("pain", "A"): dict(discard=[9]),
    ("pain", "B"): dict(split=[11]),
}


def load_S(folder):
    with h5py.File(os.path.join(folder, "final_analysis_results.mat"),
                   "r") as f:
        S3 = np.array(f["output"]["spatial_weights"])
    k, w, h = S3.shape
    return S3.transpose(0, 2, 1).reshape(k, h * w).T.astype(np.float32), (h, w)


def blobs(col, shape):
    """Partition one footprint into its blobs, largest first.

    A footprint is a weighted profile, not a mask, so each part has to keep
    the original weights - a hard-edged mask would make the least-squares
    fit worse. The first version did that by dilating each thresholded blob
    by 9x9 and copying the weights inside. That let a part reach into the
    low-weight skirt beyond the threshold, and the verification caught it:
    the two halves of pain B #11 together covered 22 px that were not in the
    original footprint at all.

    So the support is partitioned instead of grown: every nonzero pixel of
    the original goes to whichever kept blob is nearest. The parts are then
    disjoint by construction and their union is exactly the original
    support - neither half can claim the other's skirt, and nothing appears
    outside.
    """
    img = col.reshape(shape)
    bw = (img > REL * img.max()).astype(np.uint8)
    n, lab, st, _ = cv2.connectedComponentsWithStats(bw, 8)
    order = sorted((int(st[j, cv2.CC_STAT_AREA]), j)
                   for j in range(1, n))[::-1]
    order = [(a, j) for a, j in order if a >= MIN_BLOB]
    if not order:
        return []
    support = img > 0
    dists = []
    for _a, j in order:
        d = cv2.distanceTransform((lab != j).astype(np.uint8),
                                  cv2.DIST_L2, 3)
        dists.append(d)
    owner = np.argmin(np.stack(dists, 0), axis=0)
    out = []
    for idx, (a, _j) in enumerate(order):
        keep = np.where(support & (owner == idx), img, 0).astype(np.float32)
        out.append((int((keep > REL * img.max()).sum()), keep.ravel()))
    return out


def seed(cy, cx, shape, r=SEED_RADIUS):
    h, w = shape
    yy, xx = np.mgrid[0:h, 0:w]
    g = np.exp(-((yy - cy) ** 2 + (xx - cx) ** 2) / (2 * (r / 1.5) ** 2))
    g[g < 0.05] = 0
    return g.ravel().astype(np.float32)


def run(tag, plane):
    session = SESSIONS[tag]
    folder = os.path.join(session, "output_split", f"plane_{plane}")
    S, shape = load_S(folder)
    k = S.shape[1]
    d = DECISIONS.get((tag, plane), {})
    drop = set(d.get("discard", []))
    trim = set(d.get("trim", []))
    split = set(d.get("split", []))

    cols, rows = [], []
    for i in range(1, k + 1):
        if i in drop:
            rows.append(dict(label=str(i), origin=i, op="discard",
                             index=0, area_px=int((S[:, i - 1] > 0).sum())))
            continue
        if i in split:
            bl = blobs(S[:, i - 1], shape)
            if len(bl) < 2:
                raise SystemExit(f"{tag} plane {plane} #{i}: asked to split "
                                 f"but only {len(bl)} blob(s) found")
            for sfx, (a, col) in zip("abcdefg", bl):
                cols.append(col)
                rows.append(dict(label=f"{i}{sfx}", origin=i, op="split",
                                 index=len(cols), area_px=a))
            continue
        if i in trim:
            bl = blobs(S[:, i - 1], shape)
            if len(bl) < 2:
                raise SystemExit(f"{tag} plane {plane} #{i}: asked to trim "
                                 f"but only {len(bl)} blob(s) found")
            a, col = bl[0]
            cols.append(col)
            rows.append(dict(label=str(i), origin=i, op="trim (largest blob "
                                                        f"of {len(bl)})",
                             index=len(cols), area_px=a))
            continue
        cols.append(S[:, i - 1])
        rows.append(dict(label=str(i), origin=i, op="keep",
                         index=len(cols), area_px=int((S[:, i - 1] > 0).sum())))

    for name, cy, cx in d.get("add", []):
        col = seed(cy, cx, shape)
        cols.append(col)
        rows.append(dict(label=name, origin=-1,
                         op=f"add seed at row {cy:.0f} col {cx:.0f}",
                         index=len(cols), area_px=int((col > 0).sum())))

    Sc = np.stack(cols, 1) if cols else np.zeros((S.shape[0], 0), np.float32)
    # Columns are built as C-order (row-major) ravels of an (h, w) image,
    # which is numpy's default. MATLAB's reshape fills COLUMN-major, so
    # writing them as-is scrambled every footprint: reshape(S, 440, 512)
    # turned one 212 px soma into 17 stripes of 16 px spanning all 440 rows,
    # and every trace solved against it was therefore wrong. Because
    # 440 != 512 it is not even a clean transpose. So each column is
    # re-raveled in Fortran order here, once, at the boundary.
    h, w = shape
    Sc = np.stack([Sc[:, i].reshape(h, w).ravel(order="F")
                   for i in range(Sc.shape[1])], 1).astype(np.float32) \
        if Sc.shape[1] else Sc
    out = os.path.join(folder, "curated")
    os.makedirs(out, exist_ok=True)
    savemat(os.path.join(out, "curated_S.mat"),
            dict(S=Sc, fov_size=np.array(shape, float),
                 order="fortran: reshape(S(:,i), fov_size(1), fov_size(2))",
                 labels=np.array([r["label"] for r in rows if r["index"] > 0],
                                 dtype=object)))
    D = pd.DataFrame(rows)
    D.to_csv(os.path.join(out, "curated_labels.csv"), index=False)
    n_out = int((D["index"] > 0).sum())
    ops = D[D["op"] != "keep"]
    print(f"  {tag} plane {plane}: {k} -> {n_out} footprints")
    for _, r in ops.iterrows():
        print(f"      #{r['origin'] if r['origin'] > 0 else '-'} "
              f"-> {r['label']:4s}  {r['op']}  ({r['area_px']} px)")
    return n_out


def main():
    print("building curated footprint sets")
    tot = 0
    for tag in ("openfield", "pain"):
        for plane in ("A", "B"):
            tot += run(tag, plane)
    print(f"\n{tot} curated footprints across the four planes")
    print("next: apply_curation.m re-derives the traces with EXTRACT")


if __name__ == "__main__":
    main()
